src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -0,0 +1,1325 @@
+        }
+
+        [Fact]
+        public void ListPattern_NarrowedTypes()
+        {
+            var source = @"
+using System;
+class X
+{
+    static int Test(object o)
+    {
+        return o switch
+        {
+            int[] {1,2,3} => 1,
+            double[] and {1,2,3} => 2,
+            float[] [3] => 3,
+            _ => -1,
+        };
+    }
+    public static void Main()
+    {
+        Console.Write(Test(new int[] { 1, 2, 3 }));
+        Console.Write(Test(new double[] { 1, 2, 3 }));
+        Console.Write(Test(new float[] { 1, 2, 3 }));
+    } 
+}
+";
+            var compilation = CreateCompilation(source, parseOptions: TestOptions.RegularWithListPatterns, options: TestOptions.ReleaseExe);
+            compilation.VerifyEmitDiagnostics();
+            CompileAndVerify(compilation, expectedOutput: "123");
+        }
+
+        [Fact]
+        public void ListPattern_ImpossiblePattern()
+        {
+            var source = @"
+using System;
+class X
+{
+    public static void Main()
+    {
+        _ = new int[0] is [0]{1};
+        _ = new int[0] is {Length:0} and {1};
+        _ = new int[0] is [1]{}; // ok
+    } 
+}
+";
+            var compilation = CreateCompilation(source, parseOptions: TestOptions.RegularWithListPatterns, options: TestOptions.ReleaseExe);
+            compilation.VerifyEmitDiagnostics(
+                // (7,13): error CS8518: An expression of type 'int[]' can never match the provided pattern.
+                //         _ = new int[0] is [0]{1};
+                Diagnostic(ErrorCode.ERR_IsPatternImpossible, "new int[0] is [0]{1}").WithArguments("int[]").WithLocation(7, 13),
+                // (8,13): error CS8518: An expression of type 'int[]' can never match the provided pattern.
+                //         _ = new int[0] is {Length:0} and {1};
+                Diagnostic(ErrorCode.ERR_IsPatternImpossible, "new int[0] is {Length:0} and {1}").WithArguments("int[]").WithLocation(8, 13));
+        }
+
+        [Fact]
+        public void ListPattern_Symbols()
+        {
+            var source =
+@"class X
+{
+    public static void Main()
+    {
+        _ = new int[0] is [var length];
+        _ = new int[0] is {var element};
+        _ = new int[0] is {..var slice};
+    }
+}";
+            var compilation = CreateCompilation(source, parseOptions: TestOptions.RegularWithListPatterns, options: TestOptions.ReleaseExe);
+            compilation.VerifyDiagnostics();
+            var tree = compilation.SyntaxTrees[0];
+            foreach (var designation in tree.GetRoot().DescendantNodes().OfType<SingleVariableDesignationSyntax>())
+            {
+                var model = compilation.GetSemanticModel(tree);
+                var symbol = model.GetDeclaredSymbol(designation);
+                Assert.Equal(SymbolKind.Local, symbol.Kind);
+                Assert.Equal(symbol.Name == "slice" ? "int[]?" : "int", ((ILocalSymbol)symbol).Type.ToDisplayString());
```

Comment:
Why is it `int[]?` rather than `int[]` for the slice?

Reply:
I'd expected this is due to using `var`. I'll verify with an explicit type.

Reply:
Makes sense. Thanks